### PR TITLE
Added module gitlab_hook

### PR DESCRIPTION
**Pull request by dj-wasabi**
##### ISSUE TYPE

- New Module Pull Request
##### COMPONENT NAME

gitlab_hook
##### ANSIBLE VERSION

```
ansible 2.1.2.0
  config file = 
  configured module search path = Default w/o overrides
```
##### SUMMARY

This will create, update or delete a (web)hook in a project in Gitlab.

### CODE REVIEW

**Comment by dj-wasabi** [comment]
Migrated from https://github.com/ansible/ansible-modules-extras/pull/3268

File: lib/ansible/modules/source_control/gitlab_hook.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+DOCUMENTATION = '''
+---
+module: gitlab_hook
+short_description: Creates/updates/deletes hooks on Gitlab projects
```
Comment (adrianmoisey) [inline lib/ansible/modules/source_control/gitlab_hook.py:22]: This is a nitpick, I'm not 100% sure if it's right, but it looks like GitLab like to go by "GitLab" and not "Gitlab"

File: lib/ansible/modules/source_control/gitlab_hook.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+
+RETURN = '''# '''
+
+try:
+    import gitlab
+    HAS_GITLAB_PACKAGE = True
+except:
+    HAS_GITLAB_PACKAGE = False
+
+from ansible.module_utils.basic import *
```
Comment (Akasurde) [inline lib/ansible/modules/source_control/gitlab_hook.py:164]: Only import required.

File: lib/ansible/modules/source_control/gitlab_hook.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+            self._module.fail_json(msg="Failed to delete a hook: %s " % e)
+        return True
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            server_url=dict(required=True),
+            validate_certs=dict(required=False, default=True, type='bool', aliases=['verify_ssl']),
+            login_user=dict(required=False, no_log=True),
```
Comment (Akasurde) [inline lib/ansible/modules/source_control/gitlab_hook.py:287]: no need for `no_log` for login_user.

File: lib/ansible/modules/source_control/gitlab_hook.py [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+
+    # Validate some credentials configuration parameters.
+    if login_user is not None and login_password is not None:
+        use_credentials = True
+    elif login_token is not None:
+        use_credentials = False
+    else:
+        module.fail_json(msg="No login credentials are given. Use login_user with login_password, or login_token")
+
+    if login_token and login_user:
```
Comment (Akasurde) [inline lib/ansible/modules/source_control/gitlab_hook.py:336]: use `mutually_exclusive`

File: lib/ansible/modules/source_control/gitlab_hook.py [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+        if use_credentials:
+            git = gitlab.Gitlab(server_url, email=login_user, password=login_password, ssl_verify=verify_ssl)
+            git.auth()
+        else:
+            git = gitlab.Gitlab(server_url, private_token=login_token, ssl_verify=verify_ssl)
+            git.auth()
+    except Exception:
+        e = get_exception()
+        module.fail_json(msg="Failed to connect to Gitlab server: %s " % e)
+
```
Comment (Akasurde) [inline lib/ansible/modules/source_control/gitlab_hook.py:349]: Can you please add check if user is authenticated or not ? in past, there was a bug related to this, which was fixed by using [this](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/source_control/gitlab_user.py#L345-L349)

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Pull request closed without merging**